github/repos_rules.go

```diff
@@ -0,0 +1,392 @@
+func NewRequiredStatusChecksRule(params *RequiredStatusChecksRuleParameters) (rule RepositoryRule) {
+	return RepositoryRule{
+		Type:       "required_status_checks",
+		Parameters: params,
+	}
+}
+
+// NewNonFastForwardRule creates a rule as part to prevent users with push access from force pushing to matching branches.
+func NewNonFastForwardRule() (rule RepositoryRule) {
+	return RepositoryRule{
+		Type: "non_fast_forward",
+	}
+}
+
+// NewCommitMessagePatternRule creates a rule to restrict commit message patterns being pushed to matching branches.
+func NewCommitMessagePatternRule(pattern *RulePatternParameters) (rule RepositoryRule) {
+	return RepositoryRule{
+		Type:       "commit_message_pattern",
+		Parameters: pattern,
+	}
+}
+
+// NewCommitAuthorEmailPatternRule creates a rule to restrict commits with author email patterns being merged into matching branches.
+func NewCommitAuthorEmailPatternRule(pattern *RulePatternParameters) (rule RepositoryRule) {
+	return RepositoryRule{
+		Type:       "commit_author_email_pattern",
+		Parameters: pattern,
+	}
+}
+
+// NewCommitterEmailPatternRule creates a rule to restrict commits with committer email patterns being merged into matching branches.
+func NewCommitterEmailPatternRule(pattern *RulePatternParameters) (rule RepositoryRule) {
+	return RepositoryRule{
+		Type:       "committer_email_pattern",
+		Parameters: pattern,
+	}
+}
+
+// NewBranchNamePatternRule creates a rule to restrict branch patterns from being merged into matching branches.
+func NewBranchNamePatternRule(pattern *RulePatternParameters) (rule RepositoryRule) {
+	return RepositoryRule{
+		Type:       "branch_name_pattern",
+		Parameters: pattern,
+	}
+}
+
+// NewTagNamePatternRule creates a rule to restrict tag patterns contained in non-target branches from being merged into matching branches.
+func NewTagNamePatternRule(pattern *RulePatternParameters) (rule RepositoryRule) {
+	return RepositoryRule{
+		Type:       "tag_name_pattern",
+		Parameters: pattern,
+	}
+}
+
+// Ruleset represents a GitHub ruleset object.
+type Ruleset struct {
+	ID   int64  `json:"id"`
+	Name string `json:"name"`
+	// Possible values for Target are branch, tag
+	Target *string `json:"target,omitempty"`
+	// Possible values for SourceType are: Repository, Organization
+	SourceType *string `json:"source_type,omitempty"`
+	Source     string  `json:"source"`
+	// Possible values for Enforcement are: disabled, active, evaluate
+	Enforcement string `json:"enforcement"`
+	// Possible values for BypassMode are: none, repository, organization
+	BypassMode   *string            `json:"bypass_mode,omitempty"`
+	BypassActors *[]BypassActor     `json:"bypass_actors,omitempty"`
+	NodeID       *string            `json:"node_id,omitempty"`
+	Links        *RulesetLinks      `json:"_links,omitempty"`
+	Conditions   *RulesetConditions `json:"conditions,omitempty"`
+	Rules        *[]RepositoryRule  `json:"rules,omitempty"`
```

Comment:
Is `*[]RepositoryRule` necessary here?  Can't it be `[]*RepositoryRule` instead, which is preferred?